### PR TITLE
Add MDETR image backbone

### PR DESCRIPTION
**Pull request by ebsmothers**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110
* #109
* #77
* #62
* __->__ #61

This commit adds support for the ResNet-101 image backbone from MDETR.
This is basically just a wrapper around TorchVision's ResNet class
with support for fetching intermediate layers from the model so that the final pooling and
FC layers are not used. We also add the nested tensor and frozen batch norm classes from
the MDETR repo, ultimately we may factor out the nested tensor class later on.

Also added a test to confirm the results match those obtained by using MDETR's image backbone directly.

This is an initial version, will update docstrings etc if the high-level implementation looks good.

Tested with
```
$ python -m pytest -v test/modules/encoders/test_mdetr_image_encoder.py
test/modules/encoders/test_mdetr_image_encoder.py::TestMDETRImageEncoder::test_resnet_101_forward PASSED

Differential Revision: [D37390044](https://our.internmc.facebook.com/intern/diff/D37390044)

### CODE REVIEW

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+from torch import nn
+from torchmultimodal.modules.layers.normalizations import FrozenBatchNorm2d
+from torchmultimodal.utils.common import NestedTensor
+from torchvision.models._utils import IntermediateLayerGetter
+from torchvision.models.resnet import ResNet101_Weights
+
+# At this point this class is basically just a wrapper
+# around the backbone to support tensor list and IntermediateLayerGetter.
+# We should refactor to pass tensor and mask separately
+class MDETRBackbone(nn.Module):
```
Comment (RdoubleA) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:20]: is the plan to change the name of this class? Since this is a general wrapper around IntermediateLayerGetter and NestedTensor

Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:20]: Good point. To make it truly general we can probably pass return_layers as an argument, otherwise this is too specific to MDETR's ResNet implementation.

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        out = OrderedDict()
+        for name, x in xs.items():
+            mask = F.interpolate(
+                tensor_list.mask[None].float(), size=x.shape[-2:]
+            ).bool()[0]
+            out[name] = NestedTensor(x, mask)
+        return out
+
+
+def _mdetr_resnet101_backbone():
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:38]: Fair question. This is not the top-level MDETR model class, so is less likely to be imported directly. But in general I think any builder functions should probably be public, so let me change it.

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+# around the backbone to support tensor list and IntermediateLayerGetter.
+# We should refactor to pass tensor and mask separately
+class MDETRBackbone(nn.Module):
+    def __init__(self, body: nn.Module):
+        super().__init__()
+        # Note that we need this to skip pooler, flatten, and FC layers in
+        # the standard ResNet implementation.
+        self.body = IntermediateLayerGetter(body, return_layers={"layer4": 0})
+
+    def forward(self, tensor_list):
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:27]: whats the type of tensor_list

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        xs = self.body(tensor_list.tensors)
+        out = OrderedDict()
+        for name, x in xs.items():
+            mask = F.interpolate(
+                tensor_list.mask[None].float(), size=x.shape[-2:]
+            ).bool()[0]
+            out[name] = NestedTensor(x, mask)
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:34]: why is nested tensor generic thing?

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        return out
+
+
+def _mdetr_resnet101_backbone():
+    # TODO: maybe support passing arg for last dilation operation as in MDETR repo
+    body = getattr(torchvision.models, "resnet101")(
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:40]: why use getattr? cant u just import

File: torchmultimodal/modules/layers/normalizations.py [hunk 6]
```diff
@@ -20,3 +21,56 @@ def forward(self, x):
             self.eps,
         )
         return output.type_as(x)
+
+
+class FrozenBatchNorm2d(nn.Module):
```
Comment (ankitade) [inline torchmultimodal/modules/layers/normalizations.py:26]: is this very mdetr resnet specific? if yes, consider moving to that module

**Review by ankitade — COMMENTED** [review]
the input / output format are somewhat unintuitive to a lay man. where does the complexity come from?

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 4 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:34]: It's really just a wrapper around a list of tensor plus mask. I think we can factor this out later on

File: torchmultimodal/modules/layers/normalizations.py (hunk 6 above)
Comment (ebsmothers) [reply torchmultimodal/modules/layers/normalizations.py:26]: It's also used in the ResNet benchmark from FasterRCNN, for example. But not sure how widespread the usage is in general. I think moving it to the MDETR ResNet module should be reasonable

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 5 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:40]: This is just how it's done in MDETR. As far as I can tell it shouldn't make a huge difference

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 3 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:27]: It should be a `NestedTensor`. Will add the type info

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        mask = F.interpolate(image_masks[None].float(), size=tensor.shape[-2:]).bool()[
+            0
+        ]
+        return tensor, mask
+
+
+def mdetr_resnet101_backbone() -> MaskedIntermediateLayer:
+    body = resnet101(
+        replace_stride_with_dilation=[False, False, False],
+        weights=ResNet101_Weights.IMAGENET1K_V1,
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:141]: can we make this part of function args

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        ).flatten(3)
+        pos_y = torch.stack(
+            (pos_y[:, :, :, 0::2].sin(), pos_y[:, :, :, 1::2].cos()), dim=4
+        ).flatten(3)
+        pos = torch.cat((pos_y, pos_x), dim=3).permute(0, 3, 1, 2)
+        return pos
+
+
+class MaskedIntermediateLayer(nn.Module):
+    def __init__(self, body: nn.Module, intermediate_layers: Dict[str, Any]):
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:118]: isnt it Dict[str,int]?

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        assert (
+            len(intermediate_layers.keys()) == 1
+        ), "IntermediateLayerGetter for multiple layers is not supported"
+        self.body = IntermediateLayerGetter(body, return_layers=intermediate_layers)
+
+    def forward(
+        self, images: torch.Tensor, image_masks: torch.Tensor
+    ) -> Tuple[Tensor, Tensor]:
+        out = self.body(images)
+        tensor = out[next(reversed(out))]
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:131]: why do we need this?

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        mask = F.interpolate(image_masks[None].float(), size=tensor.shape[-2:]).bool()[
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:132]: some comments

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        b = self.bias.reshape(1, -1, 1, 1)  # type: ignore
+        rv = self.running_var.reshape(1, -1, 1, 1)  # type: ignore
+        rm = self.running_mean.reshape(1, -1, 1, 1)  # type: ignore
+        eps = 1e-5
+        scale = w * (rv + eps).rsqrt()
+        bias = b - rm * scale
+        return x * scale + bias
+
+
+class PositionEmbeddingSine(nn.Module):
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:70]: neither used nor tested?

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+            error_msgs,
+        )
+
+    def forward(self, x: torch.Tensor):
+        # move reshapes to the beginning to make it fuser-friendly
+        # ignore mypy errors because fixing them would break checkpoint loading
+        w = self.weight.reshape(1, -1, 1, 1)  # type: ignore
+        b = self.bias.reshape(1, -1, 1, 1)  # type: ignore
+        rv = self.running_var.reshape(1, -1, 1, 1)  # type: ignore
+        rm = self.running_mean.reshape(1, -1, 1, 1)  # type: ignore
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:63]: why not just do these in init?

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        prefix,
+        local_metadata,
+        strict,
+        missing_keys,
+        unexpected_keys,
+        error_msgs,
+    ):
+        num_batches_tracked_key = prefix + "num_batches_tracked"
+        if num_batches_tracked_key in state_dict:
+            del state_dict[num_batches_tracked_key]
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:45]: why do we need this? add comments

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+import torch
+import torch.nn.functional as F
+from torch import nn, Tensor
+from torchvision.models._utils import IntermediateLayerGetter
+from torchvision.models.resnet import resnet101, ResNet101_Weights
+
+
+class FrozenBatchNorm2d(nn.Module):
+    """
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:18]: its on your todo but add docstrings

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 8 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py]: Just refactored so this is no longer relevant. Now we only accept a single layer, since only segmentation task requires more than one intermediate layer and removing support for multiple intermediate layers simplifies the code a lot.

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 9 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:162]: Because IntermediateLayerGetter returns a dict mapping layer names to layer outputs. So even if we only pass one layer we need to extract the tensor in this way

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 10 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:163]: So I added something in the docstring, lmk if it's sufficient

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 11 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:78]: Not tested explicitly here, but its used in #77 (and tested there by loading and calling forward on the full model). In the original repo they tie this to the backbone, but I decided to keep it separate. But it is still in here as a holdover. Happy to move it elsewhere if you think that's clearer

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 14 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:18]: done

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+
+        pos_x = x_embed[:, :, :, None] / dim_t
+        pos_y = y_embed[:, :, :, None] / dim_t
+        pos_x = torch.stack(
+            (pos_x[:, :, :, 0::2].sin(), pos_x[:, :, :, 1::2].cos()), dim=4
+        ).flatten(3)
+        pos_y = torch.stack(
+            (pos_y[:, :, :, 0::2].sin(), pos_y[:, :, :, 1::2].cos()), dim=4
+        ).flatten(3)
+        pos = torch.cat((pos_y, pos_x), dim=3).permute(0, 3, 1, 2)
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:110]: shape comment

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+        num_pos_feats: int = 64,
+        temperature: int = 10000,
+        normalize: bool = False,
+        scale: float = None,
+    ):
+        super().__init__()
+        self.num_pos_feats = num_pos_feats
+        self.temperature = temperature
+        self.normalize = normalize
+        if scale is not None and normalize is False:
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:82]: can u just take in scale as optional float and do away with normalize?

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+    generalized to work on images.
+
+    Args:   num_pos_feats (int): Number of positional features
+                (should be half the output embedding size). Default = 64
+            temperature (int): Base for generating frequency mesh. Default = 10000
+            normalize (bool): Whether to normalize the image grid to values in [0,1].
+                Default = False
+            scale (float): Scaling factor when performing normalization. Default = None
+
+    Inputs: mask (Tensor): Padding mask (used to determine size of each image in batch).
```
Comment (ankitade) [inline torchmultimodal/modules/encoders/mdetr_image_encoder.py:68]: shape info?

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 11 above)
Comment (ankitade) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:55]: maybe call it 2dposition embeddings

File: torchmultimodal/modules/encoders/mdetr_image_encoder.py (hunk 15 above)
Comment (ebsmothers) [reply torchmultimodal/modules/encoders/mdetr_image_encoder.py:106]: Added in the docstring

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37390044).

**Review by ankitade — APPROVED** [review]

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37390044).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37390044).
